src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts

```diff
@@ -361,35 +363,74 @@ export class CommandDetectionCapability extends Disposable implements ICommandDe
 		this._logService.debug('CommandDetectionCapability#handleCommandStart', this._currentCommand.commandStartX, this._currentCommand.commandStartMarker?.line);
 	}
 
-	private _handleCommandStartWindows(): void {
+	private async _handleCommandStartWindows(): Promise<void> {
+		this._commandStartedWindowsBarrier = new Barrier();
 		this._currentCommand.commandStartX = this._terminal.buffer.active.cursorX;
 
 		// On Windows track all cursor movements after the command start sequence
 		this._commandMarkers.length = 0;
-		// HACK: Fire command started on the following frame on Windows to allow the cursor
-		// position to update as conpty often prints the sequence on a different line to the
-		// actual line the command started on.
-		timeout(0).then(() => {
-			if (!this._currentCommand.commandExecutedMarker) {
-				this._onCursorMoveListener = this._terminal.onCursorMove(() => {
-					if (this._commandMarkers.length === 0 || this._commandMarkers[this._commandMarkers.length - 1].line !== this._terminal.buffer.active.cursorY) {
-						const marker = this._terminal.registerMarker(0);
-						if (marker) {
-							this._commandMarkers.push(marker);
-						}
-					}
-				});
+
+		// Conpty could have the wrong cursor position at this point.
+		if (!this._cursorOnNextLine() || !this._cursorLineLooksLikeWindowsPrompt()) {
+			// Poll for 200ms until the cursor position is correct.
+			let i = 0;
+			for (; i < 20; i++) {
+				await timeout(10);
+				if (this._cursorOnNextLine() && this._cursorLineLooksLikeWindowsPrompt()) {
+					break;
+				}
 			}
-			this._currentCommand.commandStartMarker = this._terminal.registerMarker(0);
-			if (this._currentCommand.commandStartMarker) {
-				const line = this._terminal.buffer.active.getLine(this._currentCommand.commandStartMarker.line);
-				if (line) {
-					this._currentCommand.commandStartLineContent = line.translateToString(true);
+			if (i === 20) {
+				this._logService.debug('CommandDetectionCapability#_handleCommandStartWindows reached max attempts, ', this._cursorOnNextLine(), this._cursorLineLooksLikeWindowsPrompt());
+			}
+		} else {
```

Comment:
We need to cancel this process if another command happens while it's happening. It's an edge case that can maybe only reproduce in a misbehaving shell, but it could lead to really weird bugs.